### ISSUE TITLE
Simple Question: How to set the x and y-position for each TiledImage differently?

### ISSUE THREAD

**Issue by comrade2sun** [openseadragon/openseadragon#1002]
#### The idea :sparkles:

Arranging a set of images on top of a background image, giving at a first glance the overall impression of a (hopefully) artful composition.  
`updated`
→ <a href="http://homepage.o2mail.de/nixneues/OSD/utamaro_02.html">Live Example | starring Miss Wakamatsuya <strike>and her mysterious '_x' & '_y' demeanour</strike></a>

> tl;tr
> <strike>My simple ... but fruitless  _y struggle.</strike>
> Do I need a brain  the size of a universe, to solve  this case? (Definitely not!)
> Advise from a math-genius welcome. (Evermore!)
#### <strike>Beginners dream</strike> &nbsp;&nbsp;&nbsp;&nbsp;<strike>Eureka!</strike> Got it!

A simple setup like this

``` javascript
// safes 4 server request for .xml files
var chojiya={Image:{xmlns:"http://schemas" ... etc.
var araihari={Image:{xmlns:"http://schemas" ... etc.
var kagiya={Image:{xmlns:"http://schemas" ... etc.
var wakamatsu={Image:{xmlns:"http://schemas" ... etc.

var viewer = OpenSeadragon({
    id: "utamaro",
    prefixUrl: "pano/img/",
    preserveViewport: true,
    tileSources: [  
// One must imagine Sisyphus happy!
                {x: 0,     y: 0, height:0.8,   tileSource: chojiya},
                {x: 0.6,   y: 0, height:0.8,   tileSource: araihari},
                {x: 1.16,  y: 0, height:0.8,   tileSource: kagiya},
                {x: 1.76,  y: 0, height:0.8,   tileSource: wakamatsuya} // plays well with others now!
    ],
});
viewer.addHandler('open', function() {
    viewer.viewport.fitBounds(new OpenSeadragon.Rect(-0.41, -0.95, 3, 3));
});
```

<code>[Kauderwelsch deleted]</code>
I am totally aware, that 
-  <code>[Kauderwelsch deleted]</code>
- this is just a <strike>simple</strike> complicated minds <strike>dream</strike> version <strike>of an easy kind of shorthand</strike> , yet some <strike>crude mixture of </strike>home brewed JavaScript syntax and that I  attack in the completely wrong direction, but ... read next s.v.p.
#### My state of the art

Searching the comments, starring at Codepens and spending some time at the OpenSeadragon API feels like sitting on a pebble beach. It gives me the wildest ideas of what adventurous castles could be made of pebbles — <strike>just for now, I would be happy  to pick the right four or five pebbles to build my humble set of images.</strike> `Got it! Hmmm ... not really` Last but not least, what is also literally missing, any glue
 ... pardon clue <strike>to stick them together.</strike>
I have to admit, I'm an absolute JavaScript illiterate, but I'm convinced, the Seadragon <strike>will generously give me pardon</strike>  has currently much more fun as participant of  "The World Stone Skimming Championships"  at Easdale Island, ... other than reading my Kauderwelsch.

Any clue appreciated & Thanks for reading until this last sentence

p.s.
#### _Wildest future plans_

_Once I get this first "set of images" done, it's easy to use it as template and create a whole bunch of it. The future challenge will be: cycle through them back and forward, similar like in Sequence Mode. 
I am quite optimistic ..._ :sunny:

**Comment by avandecreme** [comment]
I am not sure to understand how you would like to align the images.
I made a codepen for easier collaboration here: http://codepen.io/avandecreme/pen/LkqbXG

Would you like all 4 images to have the same height?

**Comment by comrade2sun** [comment]
#### Hi avandecreme
Thank You for taking the time to write me a note.

If you don't mind, I would like to skim some more pebbles and see how far they can bounce. Have a look at this fancy stuff ... 
<a href="http://ahadaschaurija.square7.ch/rama/daydreamer.html">A Daydreamer who workes on the Graveyard Shift and sleeps all Day!</a>
... the kind of Schnickschnack I uploaded a couple of years ago to <a href="http://zoomorama.typepad.com/zooming/2010/04/zoomorama-closes-down.html">zoomorama.com (&dagger; April 29, 2010 | R.I.P.)</a>

_Please note! You may get stuck at the first attempt to load the flash player.The reason is old style embedding and code line 53. Hit F5 to reload._

You might get the idea now, what I mean with this question: "How to set the x and y-position ... etc." Just take a peek at <a href="http://ahadaschaurija.square7.ch/data/zml/daydreamer/page-1.zml">data/zml/daydreamer/page-1.zml</a>
Basically a xml file where all the magic is done. Take a look at the numbers and You will get a glimmer, why I am stuck somewhere between 'top left',  three-dimensional Euclidean space and <a href="http://homepage.o2mail.de/nixneues/OSD/utamaro_02.html">Miss Wakamatsuya :goat: gives a damn about</a>  what I am trying to achieve.

You may ask now, why all this fuss? The answer is death simple: I just want to abandon this (good) old Flash Player. Sigh!

Thanks for reading

##### _OSD unrelated musing_
_But that's not the end of all this misery. The first time in ages, I am wearing socks in the middle of august, "gefühlte 16 °C"  and the Weather Forecast claims: "Tomorrow is approximately the same temperature as today."_  
It may will rain ... it may not ... it may will Rrrrr ... :umbrella:

**Comment by iangilman** [comment]
Okay, sounds to me like you want to be able to position the tiled images not only in relation to each other but also in relation to the background image which doesn't zoom, correct?

You're on the right track for positioning them in relation to each other, by using the x and y. This doesn't help you position them in relation to the background image, however, since the group of tiled images is simply centered at the "home" position. So, you'll want to pick a specific view and tell the viewer to go there, such as:

```
viewer.viewport.setBounds(new OpenSeadragon.Rect(-3, -2, 4, 3), true);
```

You can use `preserveViewport: true` in your options so the initial opening doesn't take you back to home, overriding the setBounds you do. 

Another note is that you can use whatever coordinates you want... if it's easier to think in terms of coordinates you used in http://ahadaschaurija.square7.ch/data/zml/daydreamer/page-1.zml go for it... just remember to set the width and height of the images as well as the x and y.

Bringing it all together, it's something like:

```
// safes 4 server request for .xml files
var chojiya={Image:{xmlns:"http://schemas" ... etc.
var araihari={Image:{xmlns:"http://schemas" ... etc.
var kagiya={Image:{xmlns:"http://schemas" ... etc.
var wakamatsu={Image:{xmlns:"http://schemas" ... etc.

var viewer = OpenSeadragon({
    id: "utamaro",
    prefixUrl: "pano/img/",
    preserveViewport: true, 
    tileSources: [  
                {x: 2080, y: 591, width: 320, height: 247,   tileSource: chojiya},
                ... etc.
    ]
});

viewer.viewport.setBounds(new OpenSeadragon.Rect(2000, 500, 500, 600), true);
```

**Comment by comrade2sun** [comment]
Hi iangilman

I didn't succeed with your suggestion. In some way, the strange _x & _y behavior of the forth image ("the  Miss Wakamatsuya case") is still puzzling. At some other point, all my _x  & _y, width & height numbers are even ignored and the images are neatly and methodically horizontal <code>center center</code> arranged.

This - in the terminology of heraldry - **_mise en abîme_** rule, meant placing the elements  in the center, is the desired behavior - let's say for a librarian - who gracefully presents his medieval codices orderly in rows and columns for the first overview. A second time this rule plays a import role, if the viewer decides to have a closer look - let's say - at the page (image) in column 4 and row 2, and this page gentle shifts in <code>center center</code> position with <code>viewportHeight</code> and `Width: auto`. It's up to the viewer now, to zoomIn/Out or grab it and move the next page in sight.

My fancy ideas  want to break out of this column and row order and splatter the elements all over the background, even with different Width and Heigth! Thus the opposite from the librarians scenario. This would be for him an abyss and vice versa, this column and row order is the abyss for my fancy ideas.
 Hmmm ...  :foggy:  :wrench:

Speaking about backgrounds, indeed, the background will be always fixed and never zoomable. The backgrounds sole purpose is, to serve as stage for the other zoomable elements. This gives me a wide variety presenting images at the first sight, depending on how demanding backgrounds are created and I use them.
Like in  Miss Wakamatsuya example a more flat decorative arrangement, just to make the viewer feel (hopefully) more pleasant. In the daydreamer example with a  touch of a perspective, to offer a bit of a narrative and make the viewer (visitor) more curious in terms of  what this daydreamer is gazing at. Especially in this second scenario, part of the idea of this overall composition is, to give the visitor a fair chance to decide quickly: "Ey, this is boring sh*t - wisch&weg - or this is cula-bula, lets zoom in".
I hope you will get the point, why this "background thing" plays a role in the ideas fluffin' around my mind.

Some Questions _(just for now)_ about this code-snippet
<code>viewer.viewport.setBounds(new OpenSeadragon.Rect(-3, -2, 4, 3), true);</code>
The idea I understand quite well, but how do I interpret this snippet?
Does it mean, Dear OpenSeadragon, create a rectangel in the viewport with the aspect ratio 4 to 3 and start creating it from the left bottom corner  at the coordinates -3 on the x-axis and -2 on the y-axis (in the Cartesian coordinate system) and place the images/elements there? Where is the origin of this rectangel?
How do I "translate" this snippet in the aforementioned?
<code>viewer.viewport.setBounds(new OpenSeadragon.Rect(2000, 500, 500, 600), true);</code>

The easiest answer to all my question would be to ask OpenSeadragon: "Please be nice and paint the recangles background ionian-blue, the borders irish-green and place a fire-red dot at the origin"
Is anybody able and so kind to translate this last sentence for me (and maybe others) in OpenSeadragons dialect? The idea is, to play a bit around and understand more about OSD coordinates and boundaries.

Thanks for trying to understand my denglish explanations

**Comment by iangilman** [comment]
Yes, sounds like you need to get familiar with the viewport coordinates system. You should read this if you haven't:

http://openseadragon.github.io/examples/viewport-coordinates/

One thing to be aware of is that the origin is in the upper left, as with most computer graphics, rather than the lower left, as with most mathematics. 

This incantation (sorry, before I said `setBounds` but it's actually `fitBounds`): 

```
viewer.viewport.fitBounds(new OpenSeadragon.Rect(-3, -2, 4, 3), true);
```

...does not mean draw a rectangle on the screen, but instead means to move the viewport, which you can think of like a camera, to a specific location in the world. Imagine the viewport coordinates as an infinite grid, and with these lines:

```
                {x: -2.35,  y: -1.50,   tileSource: chojiya},
                {x: -1.30,  y: -1.60,   tileSource: araihari},
                {x: -0.25,  y: -1.45,   tileSource: kagiya},
                {x:  0.95,  y: -1.00,   tileSource: wakamatsuya} 
```

...you're laying things out in that world. What you see, and where you see it, however, has more to do with where you position the camera. If you move the camera far away from the images, you'll see nothing. If you zoom the camera in a lot, you'll only see a small portion of something, blown up to fill the screen.

If you don't take control yourself and position the camera how you want, OSD just centers the content of your scene, which isn't what you want in this case. So, you need to be friends with moving the camera, which you do with `viewer.viewport.fitBounds`. What it wants is a rectangle (not something drawn on screen, just the mathematical idea of a rectangle), which you could create with `new OpenSeadragon.Rect`.

My recommendation is to open up your page and use the browser's development console to try various `fitBounds` numbers to see what you can get.

Keep in mind that `fitBounds` uses the same coordinate system you use to lay out the images, so if you put chojiya at -2.35,-1.50 then `fitBounds` to -2.35,-1.50, you'll end up with that image appearing near the upper left of the viewer.

I say "near" because `fitBounds` will pad your rectangle if it's not the same aspect ratio as the viewer.

Anyway, see if that helps bring clarity! What you're trying to do is definitely possible.

**Comment by comrade2sun** [comment]
##### Just a brief note

<code>viewport.fitBounds</code> sounds promising and maybe this is the essential and missing point. 

Deep sighs like _"[...] could you please paint a ionian blue background ff."_, Please take them with a little twinkle in the eye. What I was _"praying"_ for, was  a temporary visual aid in the beginning. Simular like drawing the borders in the pen _"OpenSeadragon draw over with Canvas"_ or in  debug mode.
Anyway, the next step will be, spendinding some time with <a href="https://github.com/msalsbe/OpenSeadragonImagingHelper">OpenSeadragonImagingHelper</a> and another, learning more of the OSD syntaksis.

#### in nuce
Your comment was very helpful for several question and I appreciate your patience with my long winding explanations. 

Merci bien

p.s.:
I have to give it a little non OSD related break for now. The weather forecast sounds also promising :sunny: 
 ...  finally the opportunity to catch up with some other delayed activities —  means, literally off the grid and no internet for a while. But definitely, I will leave a note about my progress _(hopefully with some little shrimp for the seadragon in my basket)_ in the coming days.
So please stay tuned

**Comment by iangilman** [comment]
Sounds good... do keep me posted when you reconnect!

As for adding something to see the coordinates, you can use overlays, like so:

```
OpenSeadragon({
    ...
    overlays: [{
        x: 0.2008,
        y: 0.4778,
        width: 3,
        height: 3,
        className: 'my-overlay'
    }],
    ...
});
```

Note that those coordinates are in viewport coordinates just like the images are.

In order to see anything you'll need to style it with CSS, like so:

```
.my-overlay {
    border: 1px solid blue;
}
```

**Comment by comrade2sun** [comment]
#### Moin iangilman

Just to say hello and thanks.
Finally I had the opportunity this morning to give it a try  again. This does the trick for Miss Wakamatsuya:

```javascript
viewer.addHandler('open', function() {
    viewer.viewport.fitBounds(new OpenSeadragon.Rect(-0.41, -0.95, 3, 3));
});
```

For the other approximately 90+ panels I want to recreate and in the future, it's more or less a variation of the x and y coordinates and tuning the viewport.fitBounds as needed. Anyway, this are some good step forwards. 
Besides that, I implemented 'OSD hit detection' and it works like a charm. Some minor things I have to find out is,  getting the tiled images back to their initial position once the home button is pressed.  Playing with a custom toolbar and so on ...

The big challenge for me will be, to find a way to cycle back and forward through a bunch of different image sets, plus  presenting a kind of index page as overview of all the image sets ... means, a million more things to learn and there might be a gazillion more questions to come.
As always, I am quite optimistic ... :sunny:

**Comment by iangilman** [comment]
Great to hear!

Btw, if you want the viewer to go to position immediately rather than animating to it, you can add a second parameter to fitBounds, a `true` meaning "immediately", like so:

```
viewer.viewport.fitBounds(new OpenSeadragon.Rect(-0.41, -0.95, 3, 3), true);
```

Please share once you have something you're ready to show... I'd love to add it to:

http://openseadragon.github.io/examples/in-the-wild/

...as well as to just see it for myself!

Good luck :)

**Comment by comrade2sun** [comment]
@iangilman 
#### Enfin!
I had some spare time to proceed with OSD. The result is almost to what I wanted to achieve, as you can already see  in this kind of a 'Prototype-Image Gallery':

> <a href="http://homepage.o2mail.de/nixneues/cojp/utamaro.html">Kitagawa Utamaro</a>

For a start, this will do it, but there is still some improvement as in the code and design necessary.

Par example: 
I would like to style the `tileSource` in some cases with a `boxShadow`, in other cases with a small border,
Spending the toolbar a small loading indicator, when the tile Source is changing,
Let the tileSource(s) on load "fly in" some way,
"Fade in" the new `background-image` or some other transition 
Change background-color in fullscreen from black to `background-color` of the `body`

... or how the heck could I write a variable like this much shorter:
```javascript 
var kobikicho = {Image: {xmlns:"http://schemas.microsoft.com/deepzoom/2009",Url:"pano/kobikicho_files/",Format:"jpg",Overlap:"1",TileSize:"256",Size: {Width:"6451",Height:"9151"}}}; 
```
This part:
```javascript 
 xmlns:"http://schemas.microsoft.com/deepzoom/2009",Overlap:"1",TileSize:"256",
```
is always the same. I am sure, it's possible to store it once in a var, pass it in the other 50+ variables, get rid of a bunch of charakters and in result gain a smaller file size.
etc. pp. ... 

Speaking about "OpenSeadragon in the Wild" there is much more work to do.  Anyway, if it's all done, I wouldn't mind and definitely will leave a note.

Up till now, despite all my *"Himmel Arsch & Wolkenbruch"*  and *"What On Earth"* coding style, it was/is still a great fun. 

  Salü   🌞

**Comment by iangilman** [comment]
It's looking lovely! And of course I'm glad it's fun as well :)

Keep me posted!
